node/block_exchange.go

```diff
@@ -0,0 +1,226 @@
+	storeBatch, ok := store.(ds.Batching)
+	if !ok {
+		return nil, errors.New("failed to access the datastore")
+	}
+	ss, err := goheaderstore.NewStore[*types.SignedHeader](storeBatch)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize the header store: %w", err)
+	}
+
+	return &HeaderExchangeService{
+		conf:         conf,
+		genesis:      genesis,
+		p2p:          p2p,
+		ctx:          ctx,
+		headerStore:  ss,
+		logger:       logger,
+		syncerStatus: new(SyncerStatus),
+	}, nil
+}
+
+func (bExService *BlockExchangeService) initBlockStoreAndStartSyncer(ctx context.Context, initial *types.Block) error {
+	if err := bExService.blockStore.Init(ctx, initial); err != nil {
+		return err
+	}
+	if err := bExService.syncer.Start(bExService.ctx); err != nil {
+		return err
+	}
+	bExService.syncerStatus.m.Lock()
+	defer bExService.syncerStatus.m.Unlock()
+	bExService.syncerStatus.started = true
+	return nil
+}
+
+func (bExService *BlockExchangeService) tryInitBlockStoreAndStartSyncer(ctx context.Context, trustedBlock *types.Block) {
+	if trustedBlock != nil {
+		if err := bExService.initBlockStoreAndStartSyncer(ctx, trustedBlock); err != nil {
+			bExService.logger.Error("failed to initialize the headerstore and start syncer", "error", err)
+		}
+	}
+}
+
+func (bExService *BlockExchangeService) writeToBlockStoreAndBroadcast(ctx context.Context, signedHeader *types.SignedHeader) {
+	// For genesis header initialize the store and start the syncer
+	if signedHeader.Height() == bExService.genesis.InitialHeight {
+		if err := bExService.blockStore.Init(ctx, signedHeader); err != nil {
+			bExService.logger.Error("failed to initialize header store", "error", err)
+		}
+
+		if err := bExService.syncer.Start(bExService.ctx); err != nil {
+			bExService.logger.Error("failed to start syncer after initializing header store", "error", err)
+		}
+	}
+
+	// Broadcast for subscribers
+	if err := bExService.sub.Broadcast(ctx, signedHeader); err != nil {
+		bExService.logger.Error("failed to broadcast block header", "error", err)
+	}
+}
+
+// OnStart is a part of Service interface.
+func (bExService *BlockExchangeService) Start() error {
+	var err error
+	// have to do the initializations here to utilize the p2p node which is created on start
+	ps := bExService.p2p.PubSub()
+	bExService.sub = goheaderp2p.NewSubscriber[*types.SignedHeader](ps, pubsub.DefaultMsgIdFn, bExService.genesis.ChainID)
+	if err = bExService.sub.Start(bExService.ctx); err != nil {
+		return fmt.Errorf("error while starting subscriber: %w", err)
+	}
+	if _, err := bExService.sub.Subscribe(); err != nil {
+		return fmt.Errorf("error while subscribing: %w", err)
+	}
+
+	if err = bExService.blockStore.Start(bExService.ctx); err != nil {
+		return fmt.Errorf("error while starting header store: %w", err)
+	}
+
+	_, _, network := bExService.p2p.Info()
+	if bExService.p2pServer, err = newP2PServer(bExService.p2p.Host(), bExService.blockStore, network); err != nil {
+		return err
```

Comment:
might as well add some context to this error since you do it for every other one :-)